File: kits/js/energium/kit.js
const fs = require('fs');
const readline = require('readline');

// Create parser and use ',' as the delimiter between commands being sent by the `Match` and `MatchEngine`
const Parser = require('./parser');
const {
  GameMap
} = require('./map');
const {
  INPUT_CONSTANTS
} = require('./io');
const {
  Player,
  Base,
  Unit,
} = require('./game_objects');
const GAME_CONSTANTS = require('./game_constants');
const parse = new Parser(' ');

/**
 * Agent for sequential `Designs`
 */
class Agent {
  _setup() {

    // Prepare to read input
    const rl = readline.createInterface({
      input: process.stdin,
      output: null,
    });

    let buffer = [];
    let currentResolve;
    const makePromise = function () {
      return new Promise((resolve) => {
        currentResolve = resolve;
      });
    };
    // on each line, push line to buffer
    rl.on('line', (line) => {
      buffer.push(line);
      currentResolve();
      currentPromise = makePromise();
    });

    // The current promise for retrieving the next line
    let currentPromise = makePromise();

    // with await, we pause process until there is input
    const getLine = async () => {
      return new Promise(async (resolve) => {
        while (buffer.length === 0) {
          // pause while buffer is empty, continue if new line read
          await currentPromise;
        }
        // once buffer is not empty, resolve the most recent line in stdin, and remove it
        resolve(parse(buffer.shift()));
      });
    };
    this.getLine = getLine;
  }

  /**
   * Constructor for a new agent
   * User should edit this according to the `Design` this agent will compete under
   */
  constructor() {
    this._setup(); // DO NOT REMOVE
  }

  /**
   * Initialize Agent for the `Match`
   * User should edit this according to their `Design`
   */
  async initialize() {

    // get agent ID
    this.id = (await this.getLine()).nextInt();
    this.turn = 0;
    // get some other necessary initial input
    let mapInfo = (await this.getLine());
    let width = mapInfo.nextInt();
    let height = mapInfo.nextInt();
    this.mapWidth = width;
    this.mapHeight = height;
    this.map = new GameMap(width, height);
    this.players = [new Player(0), new Player(1)];
    await this.retrieveUpdates();
  }
  /**
   * Updates agent's own known state of `Match`
   * User should edit this according to their `Design`.
   */
  async update() {
    this.turn++;
    // wait for the engine to send any updates
    await this.retrieveUpdates();
  }

  resetPlayerStates() {
    this.players[0].units = [];
    this.players[1].units = [];
  }
  async retrieveUpdates() {
    this.resetPlayerStates();
    while (true) {
      let update = (await this.getLine());
      // console.error(update);
      if (update.str === INPUT_CONSTANTS.DONE) {
        break;
      }
      const inputIdentifier = update.nextStr();
      switch (inputIdentifier) {
        case INPUT_CONSTANTS.BASE_LOCATION: {
          const team = update.nextInt();
          const x = update.nextInt();
          const y = update.nextInt();
          this.players[team].bases.push(new Base(team, x, y));
          break;
        }
        case INPUT_CONSTANTS.TILE_POINTS: {
          const x = update.nextInt();
          const y = update.nextInt();
          const pts = update.nextInt();
          const tile = this.map.getTile(x, y);
          tile.energium = pts;
          break;
        }
        case INPUT_CONSTANTS.TEAM_POINTS: {
          const team = update.nextInt();
          const pts = update.nextInt();
          this.players[team].energium = pts;
          break;
        }
        case INPUT_CONSTANTS.UNIT_DATA: {
          const team = update.nextInt();
          const unitid = update.nextInt();
          const x = update.nextInt();
          const y = update.nextInt();
          const lastRepairTurn = update.nextInt();
          this.players[team].units.push(new Unit(team, unitid, x, y, lastRepairTurn, this.turn));
          break;
        }
      }
    }
  }

  /**
   * End a turn
   */
  endTurn() {
    console.log('D_FINISH');
  }
}

module.exports = {
  Agent
};

File: kits/python/energium/kit.py

from .game_objects import Base, Player, Unit
from .map import GameMap

def read_input():
    """
    Reads input from stdin
    """
    try:
        return input()
    except EOFError as eof:
        raise SystemExit(eof)
      
class Agent:
    """
    Constructor for a new agent
    User should edit this according to their `Design`
    """
    def __init__(self):
        pass

    """
    Initialize Agent for the `Match`
    User should edit this according to their `Design`
    """
    def initialize(self):
        self.id = int(read_input())
        self.turn: int = 0
        mapInfo = read_input().split(" ")
        width = int(mapInfo[0])
        height = int(mapInfo[1])
        self.mapWidth = width;
        self.mapHeight = height;
        self.map = GameMap(width, height);
        self.players: list[Player] = [Player(0), Player(1)];
        self.retrieve_updates()

    def resetPlayerStates(self):
        self.players[0].units = [];
        self.players[1].units = [];

    def retrieve_updates(self):
        self.resetPlayerStates()
        while(True):
            str = read_input()
            if str == "D_DONE":
                break
            update = str.split(" ")
            input_id = update[0]
            if input_id == 'b':
                team = int(update[1])
                x = int(update[2])
                y = int(update[3])
                self.players[team].bases.append(Base(team, x, y));
            elif input_id == 't':
                x = int(update[1])
                y = int(update[2])
                pts = int(update[3])
                tile = self.map.get_tile(x, y);
                tile.energium = pts;
            elif input_id == 'p':
                team = int(update[1])
                pts = int(update[2])
                self.players[team].energium = pts
            elif input_id == 'u':
                team = int(update[1])
                unitid = int(update[2])
                x = int(update[3])
                y = int(update[4])
                lrt = int(update[5])
                self.players[team].units.append(Unit(team, unitid, x, y, lrt, self.turn));



    """
    Updates Agent's own known state of `Match`
    User should edit this according to their `Design
    """
    def update(self):
        self.turn += 1
        self.retrieve_updates()

    """
    End a turn
    """
    def end_turn(self):
        print('D_FINISH')
        
